### PR TITLE
[Serve] Add max_replicas_per_node support

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?

Add a max_replicas_per_node replica config: this allows users to config how many replicas to lose for a node failure.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/http_state.py [hunk 1]
```diff
@@ -122,13 +122,17 @@ def update_actor_details(self, **kwargs) -> None:
...
         if self._health_check_obj_ref:
             finished, _ = ray.wait([self._health_check_obj_ref], timeout=0)
             if finished:
                 self._health_check_obj_ref = None
                 try:
                     ray.get(finished[0])
-                    self.try_update_status(HTTPProxyStatus.HEALTHY)
+                    # Call to reset _consecutive_health_check_failures
+                    # the status should be unchanged.
+                    self.try_update_status(self._status)
```
Comment (jjyao) [inline python/ray/serve/_private/http_state.py:135]: This is the bug I found during test: if a node is draining, we shouldn't set the status back to healthy even if health check succeeds.

File: python/ray/serve/_private/deployment_scheduler.py [hunk 2]
```diff
@@ -219,9 +221,19 @@ def _schedule_spread_deployment(self, deployment_name: str) -> None:
...
+            actor_options = copy.copy(replica_scheduling_request.actor_options)
+            if replica_scheduling_request.max_replicas_per_node is not None:
+                if "resources" not in actor_options:
+                    actor_options["resources"] = {}
+                # Using implicit resource (resources that every node
+                # implicitly has and total is 1)
+                # to limit the number of replicas on a single node.
+                actor_options["resources"][
+                    f"{ray._raylet.IMPLICIT_RESOURCE_PREFIX}{deployment_name}"
+                ] = (1.0 / replica_scheduling_request.max_replicas_per_node)
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_scheduler.py:233]: Do we need to worry about any float rounding issues here when casting to C++?

File: python/ray/serve/api.py [hunk 3]
```diff
@@ -319,6 +320,8 @@ class MyDeployment:
             shut down before being forcefully killed. Defaults to 20s.
         is_driver_deployment: [EXPERIMENTAL] when set, exactly one replica of this
             deployment runs on every node (like a daemon set).
+        max_replicas_per_node: The max number of deployment replicas can
+            run on a single node. Defaults to no limitation.
```
Comment (edoakes) [inline python/ray/serve/api.py:324]: mark this as [EXPERIMENTAL] please

File: python/ray/serve/schema.py [hunk 4]
```diff
@@ -250,6 +250,14 @@ class DeploymentSchema(
         ),
     )
 
+    max_replicas_per_node: int = Field(
+        default=DEFAULT.VALUE,
+        description=(
+            "The max number of deployment replicas can run on a single node. "
+            "Defaults to no limitation."
```
Comment (edoakes) [inline python/ray/serve/schema.py:257]: ditto mark experimental

(I wish these were unified)

**Review by edoakes — COMMENTED** [review]
Looks great, just a few nits

Thanks again for the replica config change, simplifies this a lot

**Review by edoakes — APPROVED** [review]

**Pull request merged**